go/libraries/doltcore/sqle/dfunctions/dolt_pull.go

```diff
@@ -0,0 +1,200 @@
+	return NewPullFunc(ctx, children...)
+}
+
+func (d DoltPullFunc) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	dbName := ctx.GetCurrentDatabase()
+
+	if len(dbName) == 0 {
+		return noConflicts, fmt.Errorf("empty database name.")
+	}
+
+	sess := dsess.DSessFromSess(ctx.Session)
+	dbData, ok := sess.GetDbData(ctx, dbName)
+
+	if !ok {
+		return noConflicts, fmt.Errorf("Could not load database %s", dbName)
+	}
+
+	ap := cli.CreateMergeArgParser()
+	args, err := getDoltArgs(ctx, row, d.Children())
+
+	apr, err := ap.Parse(args)
+	if err != nil {
+		return noConflicts, err
+	}
+
+	if apr.NArg() > 1 {
+		return noConflicts, actions.ErrInvalidPullArgs
+	}
+
+	var remoteName string
+	if apr.NArg() == 1 {
+		remoteName = apr.Arg(0)
+	}
+
+	pullSpec, err := env.ParsePullSpec(ctx, dbData.Rsr, remoteName, apr.Contains(cli.SquashParam), apr.Contains(cli.NoFFParam), apr.Contains(cli.ForceFlag))
+	if err != nil {
+		return noConflicts, err
+	}
+
+	srcDB, err := pullSpec.Remote.GetRemoteDBWithoutCaching(ctx, dbData.Ddb.ValueReadWriter().Format())
+	if err != nil {
+		return noConflicts, fmt.Errorf("failed to get remote db; %w", err)
+	}
+
+	ws, err := sess.WorkingSet(ctx, dbName)
+	if err != nil {
+		return noConflicts, err
+	}
+
+	var conflicts interface{}
+	for _, refSpec := range pullSpec.RefSpecs {
+		remoteTrackRef := refSpec.DestRef(pullSpec.Branch)
+
+		if remoteTrackRef != nil {
+
+			// todo: can we pass nil for either of the channels?
+			srcDBCommit, err := actions.FetchRemoteBranch(ctx, dbData.Rsw.TempTableFilesDir(), pullSpec.Remote, srcDB, dbData.Ddb, pullSpec.Branch, remoteTrackRef, runProgFuncs, stopProgFuncs)
+			if err != nil {
+				return noConflicts, err
+			}
+
+			// TODO: this could be replaced with a canFF check to test for error
+			err = dbData.Ddb.FastForward(ctx, remoteTrackRef, srcDBCommit)
+			if err != nil {
+				return noConflicts, fmt.Errorf("fetch failed; %w", err)
+			}
+
+			roots, ok := sess.GetRoots(ctx, dbName)
+			if !ok {
+				return noConflicts, sql.ErrDatabaseNotFound.New(dbName)
+			}
+
+			mergeSpec, err := sqlMergeSpec(ctx, sess, dbName, apr, remoteTrackRef.String())
+			if err != nil {
+				return noConflicts, err
+			}
+			ws, conflicts, err = mergeHelper(ctx, sess, roots, ws, dbName, mergeSpec)
+			if !errors.Is(doltdb.ErrUpToDate, err) {
+				return conflicts, err
+			}
+
+			err = sess.SetWorkingSet(ctx, dbName, ws, nil)
+			if err != nil {
+				return conflicts, err
+			}
+		}
+	}
+
+	err = actions.FetchFollowTags(ctx, dbData.Rsw.TempTableFilesDir(), srcDB, dbData.Ddb, runProgFuncs, stopProgFuncs)
+	if err != nil {
+		return noConflicts, err
+	}
+
+	return noConflicts, nil
+}
+
+func pullerProgFunc(pullerEventCh chan datas.PullerEvent) {
+	for _ = range pullerEventCh {
+	}
+}
+
+func progFunc(progChan chan datas.PullProgress) {
+	done := false
+	for !done {
+		select {
+		case _, ok := <-progChan:
+			if !ok {
+				done = true
+			}
+		case <-time.After(250 * time.Millisecond):
```

Comment:
You probably don't want this. If the pull slows down enough, you'll stop reading from the prog channel, which means it could fill up and additional sends could block.

If you want to do this right, then you need to include the *sql.Context in your select here, and check to see if it's been canceled.

Or you could just do a range on the progChan and don't do any timeout logic, that's fine for a first pass.